cpp/convert_json_to_marc.cc

```diff
@@ -0,0 +1,493 @@
+      field_contents_prefix_(field_contents_prefix), required_(required)
+{
+    if (not overflow_tag_.empty() and repeat_field_)
+        LOG_ERROR("field \"" + name_ + "\" can't have both, an over flow tag and being a repeat field!");
+
+    if (subfield_codes_to_json_tags_.empty() and json_tag_.empty())
+        LOG_ERROR("field \"" + name_
+                  + "\" neither a mapping to the contents of a control field nor the contents of data subfields has been specified!");
+}
+
+
+class JSONNodeToBibliographicLevelMapper {
+    std::string json_tag_;
+    MARC::Record::BibliographicLevel default_;
+    std::vector<std::pair<RegexMatcher *, MARC::Record::BibliographicLevel>> regex_to_bibliographic_level_map;
+public:
+    JSONNodeToBibliographicLevelMapper(const std::string &item_type_tag, const std::string &item_type_map);
+    MARC::Record::BibliographicLevel getBibliographicLevel(const JSON::ObjectNode &object_node) const;
+    MARC::Record::BibliographicLevel getBibliographicLevel(const std::string &string_value) const;
+};
+
+
+// Expects two values separated by a colon.  Backslash-escaped colons do not count as the separator colon!
+void SplitPatternAndItemType(const std::string &pattern_and_type, std::string * const pattern, std::string * const type) {
+    bool escaped(false), in_pattern(true);
+    for (const char ch : pattern_and_type) {
+        if (escaped) {
+            if (in_pattern)
+                *pattern += ch;
+            else
+                *type += ch;
+            escaped = false;
+        } else if (ch == '\\')
+            escaped = true;
+        else if (ch == ':') {
+            if (not in_pattern)
+                LOG_ERROR("bad regex pattern to item type mapping with additional unescaped colon: \"" + pattern_and_type + "\"!");
+            in_pattern = false;
+        } else {
+            if (in_pattern)
+                *pattern += ch;
+            else
+                *type += ch;
+            escaped = false;
+        }
+    }
+
+    if (escaped)
+        LOG_ERROR("bad regex pattern to item type mapping: \"" + pattern_and_type + "\"!");
+}
+
+
+MARC::Record::BibliographicLevel MapTypeStringToBibliographicLevel(const std::string &item_type) {
+    if (item_type == "monograph")
+        return MARC::Record::MONOGRAPH_OR_ITEM;
+    else if (item_type == "book chapter")
+        return MARC::Record::MONOGRAPHIC_COMPONENT_PART;
+    else if (item_type == "journal article")
+        return MARC::Record::SERIAL_COMPONENT_PART;
+    else
+        LOG_ERROR("\"" + item_type + "\" is not a valid item type!");
+}
+
+
+JSONNodeToBibliographicLevelMapper::JSONNodeToBibliographicLevelMapper(const std::string &item_type_tag, const std::string &item_type_map)
+    : json_tag_(item_type_tag)
+{
+    default_ = MARC::Record::UNDEFINED;
+
+    std::vector<std::string> pattens_and_types;
+    StringUtil::Split(item_type_map, '|', &pattens_and_types);
```

Comment:
The pipe character is used in regular expressions. You'll need to find another delimiter or make sure that it doesn't split inside patterns.

Reply:
I now support backslash escapes for colons and vertical bars.